pkg/addonmanager/controllers/addoninstall/controller.go

```diff
@@ -81,36 +81,29 @@ func (c *addonInstallController) sync(ctx context.Context, syncCtx factory.SyncC
 		return nil
 	}
 
+	var errs []error
+
 	for addonName, addon := range c.agentAddons {
 		if addon.GetAgentAddonOptions().InstallStrategy == nil {
 			continue
 		}
 
-		switch addon.GetAgentAddonOptions().InstallStrategy.Type {
-		case agent.InstallAll:
-			return c.applyAddon(ctx, addonName, clusterName, addon.GetAgentAddonOptions().InstallStrategy.InstallNamespace)
-		case agent.InstallByLabel:
-			labelSelector := addon.GetAgentAddonOptions().InstallStrategy.LabelSelector
-			if labelSelector == nil {
-				klog.Warningf("installByLabel strategy is set, but label selector is not set")
-				return nil
-			}
-
-			selector, err := metav1.LabelSelectorAsSelector(labelSelector)
-			if err != nil {
-				klog.Warningf("labels selector is not correct: %v", err)
-				return nil
-			}
-
-			if !selector.Matches(labels.Set(cluster.Labels)) {
-				return nil
-			}
+		managedClusterFilter := addon.GetAgentAddonOptions().InstallStrategy.GetManagedClusterFilter()
+		if managedClusterFilter == nil {
+			continue
+		}
+		if !managedClusterFilter(cluster) {
+			klog.Infof("managed cluster fileter is not match for addon %s on %s", addonName, clusterName)
+			continue
+		}
 
-			return c.applyAddon(ctx, addonName, clusterName, addon.GetAgentAddonOptions().InstallStrategy.InstallNamespace)
+		err = c.applyAddon(ctx, addonName, clusterName, addon.GetAgentAddonOptions().InstallStrategy.InstallNamespace)
```

Comment:
the c.applyAddon func was called 3 times in every case section, is it better to merge?

Reply:
Done, please take anothr view.
But it also cause 1 test case not appropriate, please also take a look at the test. Thanks!